File: quickstart_guides/linked_lists/java/RemoveUnsortedDuplicates.java
/*
 * Title: Remove unsorted duplicates
 *
 * Problem:
 *   Write a removeDuplicates() function which takes a list and deletes any
 *   duplicate nodes from the list. The list is not sorted.
 *
 *   For example if the linked list is 12->11->12->21->41->43->21 then
 *   removeDuplicates() should convert the list to 12->11->21->41->43.
 *
 * Execution: javac RemoveUnsortedDuplicates.java && java RemoveUnsortedDuplicates
 */


class RemoveUnsortedDuplicates {

    static Node head;

    static class Node {

        int data;
        Node next;

        Node(int d) {
            data = d;
            next = null;
        }
    }

    /* Function to remove duplicates from an
       unsorted linked list */
    void remove_duplicates() {
        Node ptr1 = null, ptr2 = null, dup = null;
        ptr1 = head;

        /* Pick elements one by one */
        while (ptr1 != null && ptr1.next != null) {
            ptr2 = ptr1;

            /* Compare the picked element with rest
                of the elements */
            while (ptr2.next != null) {

                /* If duplicate then delete it */
                if (ptr1.data == ptr2.next.data) {

                    /* sequence of steps is important here */
                    dup = ptr2.next;
                    ptr2.next = ptr2.next.next;
                    System.gc();
                } else /* This is tricky */ {
                    ptr2 = ptr2.next;
                }
            }
            ptr1 = ptr1.next;
        }
    }

    void printList(Node node) {
        while (node != null) {
            System.out.print(node.data + " ");
            node = node.next;
        }
    }

    public static void main(String[] args) {
        RemoveUnsortedDuplicates list = new RemoveUnsortedDuplicates();
        list.head = new Node(10);
        list.head.next = new Node(12);
        list.head.next.next = new Node(11);
        list.head.next.next.next = new Node(11);
        list.head.next.next.next.next = new Node(12);
        list.head.next.next.next.next.next = new Node(11);
        list.head.next.next.next.next.next.next = new Node(10);

        System.out.println("Linked List before removing duplicates : \n ");
        list.printList(head);

        list.remove_duplicates();
        System.out.println("");
        System.out.println("Linked List after removing duplicates : \n ");
        list.printList(head);
    }
}

File: quickstart_guides/linked_lists/python/remove_unsorted_duplicates.py
"""
Title: Linked list remove unsorted duplicates

Problem:
    Removes duplicates in an unsorted linked list.

Execution: python remove_unsorted_duplicates.py
"""
from typing import List
import unittest


class Node:
    """Linked list Node class."""

    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    """Generic linked list class."""

    def __init__(self):
        self.head = None

    def print_list(self):
        """Print linked list."""
        cur_node = self.head
        while cur_node:
            print(cur_node.data)
            cur_node = cur_node.next

    def append(self, data):
        """Append node to end of linked list."""
        new_node = Node(data)

        if self.head is None:
            self.head = new_node
            return

        last_node = self.head
        while last_node.next:
            last_node = last_node.next
        last_node.next = new_node

    def remove_duplicates(self):
        """Remove duplicates from an (unsorted) linked list."""
        cur = self.head
        prev = None

        dup_values = dict()

        while cur:
            if cur.data in dup_values:
                # Remove node:
                prev.next = cur.next
            else:
                # Have not encountered element before.
                dup_values[cur.data] = 1
                prev = cur
            cur = prev.next


class TestRemoveDuplicates(unittest.TestCase):
    """Unit test for remove_unsorted_duplicates."""

    def test_1(self):
        llist = LinkedList()
        llist.append(1)
        llist.append(6)
        llist.append(1)
        llist.append(4)
        llist.append(2)
        llist.append(2)
        llist.append(4)

        llist.remove_duplicates()

        cur_node = llist.head
        nodes: List[str] = []
        while cur_node:
            nodes.append(cur_node.data)
            cur_node = cur_node.next
        self.assertEqual(nodes, [1, 6, 4, 2])

    def test_2(self):
        llist = LinkedList()
        llist.append(1)
        llist.append(2)
        llist.append(2)
        llist.append(1)
        llist.append(2)
        llist.append(2)
        llist.append(1)

        llist.remove_duplicates()

        cur_node = llist.head
        nodes: List[str] = []
        while cur_node:
            nodes.append(cur_node.data)
            cur_node = cur_node.next
        self.assertEqual(nodes, [1, 2])

    def test_3(self):
        llist = LinkedList()
        llist.append(3)
        llist.append(2)
        llist.append(1)

        llist.remove_duplicates()

        cur_node = llist.head
        nodes: List[str] = []
        while cur_node:
            nodes.append(cur_node.data)
            cur_node = cur_node.next
        self.assertEqual(nodes, [3, 2, 1])


if __name__ == "__main__":
    unittest.main()
